### ISSUE TITLE
Problem managing arrays.

### ISSUE THREAD

**Issue by RobJmz** [diegohaz/rest#121]
I got problems trying to save Arrays on the database... Although the server response I receive is my modified Object...

Let me explain a bit:
First Using the rest generator I created a bunch of properties, and every one got created as "String".
I modified them, and my target schema is the following:

(It's bigger but this is the focusing property)
```
const userSchema = new Schema({
  wroups:[{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Wroup'
  }]
```

After that I modified my index.js:
```
router.put('/:id',
  token({ required: true }),
  body({ name, picture, alias, announcements, description, chats, colorDialog, colorText, contacts, dateBirth, lastname, location, notes, pinnedMessagesChats, pinnedMessagesWroups, tasks, username, userBanned, wroups }),
  update)

```
After that, the only extra code I added was to controllers (update) (For debugging to see what was happening)
```
export const update = ({ bodymen: { body }, params, user }, res, next) =>
  User.findById(params.id === 'me' ? user.id : params.id)
    .then(notFound(res))
    .then((result) => {
      if (!result) return null
      const isAdmin = user.role === 'admin'
      const isSelfUpdate = user.id === result.id
      if (!isSelfUpdate && !isAdmin) {
        res.status(401).json({
          valid: false,
          message: 'You can\'t change other user\'s data'
        })
        return null
      }
      return result
    })
<<<<<MY ADDED CODE>>>>>>
    .then((user) => {
      if(user){
        console.log(user);
        console.log(body);
        return user
      }else{
        return null
      }
    })
<<<<<MY ADDED CODE>>>>>>
    .then((user) => user ? _.merge(user, body).save() : null)
    .then((user) => user ? user.view(true) : null)
    .then(success(res))
    .catch(next)

```

My post is the following (I am using vue.js)
```
this.$http.post('https://wroups.herokuapp.com/auth', { access_token: this.user.access_token }, options).then(response => {
        // success callback
        this.user = response.body.user;
        this.user.access_token = response.body.token;
        this.user.wroups.push('59f2cb797b0d780012041be8');
        this.user.wroups.push('59f2cb797b0d780012041be4');
        this.user.alias = 'Rob';
        var options = {
          headers: {
            'Content-Type': 'application/json'
          }
        };

        this.$http.put('https://wroups.herokuapp.com/users/' + this.user.id, this.user, options).then(response => {
          console.log(this.user);
        }, response => {
          // error callback
        });
      }, response => {
        // error callback
      });
```

My alias changes perfectly.
The wroups property doesn't change.

I found out that using the following WILL CHANGE the property:
`this.user.wroups = '59f2cb797b0d780012041be4';
`
But I am using an array, not a single objectId....

When debbuging the server I found that I am not receiving any arrays... so _.merge(user, body).save() is indeed not saving anything....

![image](https://user-images.githubusercontent.com/10425977/32403570-0de25a9e-c103-11e7-94c0-ad0fcb9d5d44.png)

Any clue what this might be?

Thanks!

**Comment by maustand** [comment]
Hi

Have you tried to define all your properties via bodymen ? Just like this :

```
router.put('/:id',
  token({ required: true }),
  body({
  	name: {
  		type: String,
      	required: true
    },
    picture: {
  		type: String,
      	required: true
    },
//.......... resting propertyes here.....
    wroups: {
  		type: [String], // Array Typed
      	required: true
    }),
myUpdate

```

Currently i am using bodymen and it works perfect when i want to get an array typed property.

Let me know if it works for you..

Regards

**Comment by RobJmz** [comment]
A few hours after I Sent you the email I defined them. Object, Object arrays work perfectly. Problem came with ObjectIds. Tried to set them as String array, Object arrays, but they are not saving. How should an array of ObjectIds should be Sent?

Should I use the string array?

Thanks for your reply!
Rob

On Nov 5, 2017 4:48 AM, Mauricio Stand <notifications@github.com> wrote:

Hi

Have you tried to define all your properties via bodymen ? Just like this :

router.put('/:id',
  token({ required: true }),
  body({
        name: {
                type: String,
        required: true
    },
    picture: {
                type: String,
        required: true
    },
//.......... resting propertyes here.....
    wroups: {
                type: [String], // Array Typed
        required: true
    }),
myUpdate



Currently i am using bodymen and it works perfect when i want to get an array typed property.

Let me know if it works for you..

Regards

**Comment by maustand** [comment]
But your problem is  related to you are not getting the parameter value and it is not about saving..

Honestly i never tried with "[String]" as parameter type, but it should work.
I have been sending [Number] and computed objets and it works perfect.

Remember there is not ObjectID type in JavaScript and if you send a array of strings, it should work.

Could you send me a examples of your method  definition (index.js ) ?


Anyway if you still have problem with this, i advice you to use object.. Bodymen does not have deep type checking, so it means that it would not run a loop for checking if every element of the array is a string..

If you desire to perform a deep type checking, i advice you to use a middleware method.

Regards

**Comment by RobJmz** [comment]
Sorry I was not able to answer yesterday, here is my index.js:

```
import { Router } from 'express'
import { middleware as query } from 'querymen'
import { middleware as body } from 'bodymen'
import { password as passwordAuth, master, token } from '../../services/passport'
import { index, showMe, show, findByEmail, create, update, updatePassword, destroy } from './controller'
import { schema } from './model'
export User, { schema } from './model'

const router = new Router()
const { email, password, name, picture, description, role, alias, announcements, chats, colorDialog, colorText, contacts, dateBirth, lastname, location, notes, pinnedMessagesChats, pinnedMessagesWroups, tasks, username, userBanned, wroups } = schema.tree

/**
 * @api {get} /users Retrieve users
 * @apiName RetrieveUsers
 * @apiGroup User
 * @apiPermission admin
 * @apiParam {String} access_token User access_token.
 * @apiUse listParams
 * @apiSuccess {Object[]} users List of users.
 * @apiError {Object} 400 Some parameters may contain invalid values.
 * @apiError 401 Admin access only.
 */
router.get('/',
  token({ required: true, roles: ['admin'] }),
  query(),
  index)

/**
 * @api {get} /users/me Retrieve current user
 * @apiName RetrieveCurrentUser
 * @apiGroup User
 * @apiPermission user
 * @apiParam {String} access_token User access_token.
 * @apiSuccess {Object} user User's data.
 */
router.get('/me',
  token({ required: true }),
  showMe)

/**
 * @api {get} /users/:id Retrieve user
 * @apiName RetrieveUser
 * @apiGroup User
 * @apiPermission public
 * @apiSuccess {Object} user User's data.
 * @apiError 404 User not found.
 */
router.get('/:id',
  show)

/**
 * @api {get} /users/email/:email Retrieve user by Email
 * @apiName RetrieveUserByEmail
 * @apiGroup User
 * @apiPermission public
 * @apiSuccess {Object} user User's data.
 * @apiError 404 User not found.
 */
router.get('/email/:email',
  findByEmail)

/**
 * @api {post} /users Create user
 * @apiName CreateUser
 * @apiGroup User
 * @apiPermission master
 * @apiParam {String} access_token Master access_token.
 * @apiParam {String} email User's email.
 * @apiParam {String{8..}} password User's password.
 * @apiParam {String} [name] User's name.
 * @apiParam {String} [picture] User's picture.
 * @apiParam {String} alias User's alias.
 * @apiParam {Array[ObjectId]=Announcement} [announcements] User's announcements.
 * @apiParam {Array[ObjectId]=Chat} [chats] User's chats.
 * @apiParam {String} colorDialog=#ffffff User's color for dialogs in hex value.
 * @apiParam {String} colorText=#000000 User's color for text inside dialogs in hex value.
 * @apiParam {Array[ObjectId]=User} [contacts] User's contacts.
 * @apiParam {Date} dateBirth User's birthdate.
 * @apiParam {String{..180}} [description] User's description.
 * @apiParam {Array[ObjectId]=Events} [events] User's events.
 * @apiParam {String} lastname User's lastname.
 * @apiParam {Object=coordinates[lat(String),lng(String)]} [location] User's location coordinates.
 * @apiParam {Array[Object]=dateCreated(Date),message(String),title(String)} [notes] User's notes.
 * @apiParam {Array[ObjectId]=MessageChat} [pinnedMessagesChats] User's chats pinned messages.
 * @apiParam {Array[ObjectId]=MessageWroup} [pinnedMessagesWroups] User's wroups pinned messages.
 * @apiParam {String} [username] User's username.
 * @apiParam {Array[ObjectId]=User} [userBanned] User's banned contacts.
 * @apiParam {Array[ObjectId]=Wroup} [wroups] User's wroups.
 * @apiParam {String=user,admin} [role=user] User's role.
 * @apiSuccess (Sucess 201) {Object} user User's data.
 * @apiError {Object} 400 Some parameters may contain invalid values.
 * @apiError 401 Master access only.
 * @apiError 409 Email already registered.
 */
router.post('/',
  master(),
  body({ email, password, name, picture, description, role, alias, announcements, chats, colorDialog, colorText, contacts, dateBirth, lastname, location, notes, pinnedMessagesChats, pinnedMessagesWroups, tasks, username, userBanned, wroups }),
  create)

/**
 * @api {put} /users/:id Update user
 * @apiName UpdateUser
 * @apiGroup User
 * @apiPermission user
 * @apiParam {String} access_token Master access_token.
 * @apiParam {String} [name] User's name.
 * @apiParam {String} [picture] User's picture.
 * @apiParam {String} alias User's alias.
 * @apiParam {Array[ObjectId]=Announcement} [announcements] User's announcements.
 * @apiParam {Array[ObjectId]=Chat} [chats] User's chats.
 * @apiParam {String} colorDialog=#ffffff User's color for dialogs in hex value.
 * @apiParam {String} colorText=#000000 User's color for text inside dialogs in hex value.
 * @apiParam {Array[ObjectId]=User} [contacts] User's contacts.
 * @apiParam {Date} dateBirth User's birthdate.
 * @apiParam {String{..180}} [description] User's description.
 * @apiParam {Array[ObjectId]=Events} [events] User's events.
 * @apiParam {String} lastname User's lastname.
 * @apiParam {Object=coordinates[lat(String),lng(String)]} [location] User's location coordinates.
 * @apiParam {Array[Object]=dateCreated(Date),message(String),title(String)} [notes] User's notes.
 * @apiParam {Array[ObjectId]=MessageChat} [pinnedMessagesChats] User's chats pinned messages.
 * @apiParam {Array[ObjectId]=MessageWroup} [pinnedMessagesWroups] User's wroups pinned messages.
 * @apiParam {String} [username] User's username.
 * @apiParam {Array[ObjectId]=User} [userBanned] User's banned contacts.
 * @apiParam {Array[ObjectId]=Wroup} [wroups] User's wroups.
 * @apiParam {String=user,admin} [role=user] User's role.
 * @apiSuccess {Object} user User's data.
 * @apiError {Object} 400 Some parameters may contain invalid values.
 * @apiError 401 Current user or admin access only.
 * @apiError 404 User not found.
 */
router.put('/:id',
  token({ required: true }),
  body({ 
    name, 
    picture, 
    alias, 
    announcements, 
    description, 
    chats, 
    colorDialog, 
    colorText, 
    contacts, 
    dateBirth, 
    lastname, 
    location: Object, 
    notes: [Object], 
    pinnedMessagesChats, 
    pinnedMessagesWroups, 
    username,
    userBanned, 
    wroups: [String]}),
  update)

/**
 * @api {put} /users/:id/password Update password
 * @apiName UpdatePassword
 * @apiGroup User
 * @apiHeader {String} Authorization Basic authorization with email and password.
 * @apiParam {String{6..}} password User's new password.
 * @apiSuccess (Success 201) {Object} user User's data.
 * @apiError {Object} 400 Some parameters may contain invalid values.
 * @apiError 401 Current user access only.
 * @apiError 404 User not found.
 */
router.put('/:id/password',
  passwordAuth(),
  body({ password }),
  updatePassword)

/**
 * @api {delete} /users/:id Delete user
 * @apiName DeleteUser
 * @apiGroup User
 * @apiPermission admin
 * @apiParam {String} access_token User access_token.
 * @apiSuccess (Success 204) 204 No Content.
 * @apiError 401 Admin access only.
 * @apiError 404 User not found.
 */
router.delete('/:id',
  token({ required: true, roles: ['admin'] }),
  destroy)

export default router
```
As you can see, I set up: 
    wroups: [String]

In the default update method it goes like this:
_.merge(user, body).save()

I printed out _.merge(user,body) and it shows:
wroups: [ '59f2cb797b0d780012041be8', '59f2cb797b0d780012041b77' ],
So apparently the strings are reaching now

But they are not saving in the database...
I also tried doing it like this:
.```
then((user) => {
      if(user){
        console.log(_.merge(user, body));
        user = _.merge(user, body);
        console.log(user);
        user.save()
        return user
      }else{
        return null
      }
    })
    //.then((user) => user ? _.merge(user, body).save() : null)
```

Commenting the default method and adding up my own, and same problem...

My schema is defined like this:
```
  wroups:[{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Wroup'
  }]
```

And the server is responding with:
 ```
   "wroups": [
        "59f2cb797b0d780012041be8",
        "59f2cb797b0d780012041b77"
    ]
```

I can't seem to find out why if server responds, I got the strings on my update method now ( before I didn't, after I set up [String] on the bodymen they appeared), and they are valid ObjectIds, they still aren't saving, any clue?

Also: If I update any other field type (Strings, Arrays, Objects, Dates) they get updated properly...

**Comment by maustand** [comment]
@RobJmz 

1. could you edit your last comment and change the index.js block  ? it would be great to be on focus 
with the declaration of your service that it is giving you problems.. u need not publish  the whole index.js.

2. If the data is arriving to your controller method and you can see it in console, it means that it is not a problem with the generator and bodymen asit used to . Probably you have a problem with the way as you are trying to save the data..

3. As i was able to understand your code, you have multiples problems,
For example, you are trying to save the document using this way :

```
.then((user) => {
      if(user){
        console.log(_.merge(user, body));
        user = _.merge(user, body);
        console.log(user);
        user.save()
        return user
      }else{
        return null
      }
    })
    //.then((user) => user ? _.merge(user, body).save() : null)
```

If you dont know it, Nodejs works asynchronically ( this includes mongoosejs ), so this means that every function  that you want to call for performing something ( like this case, save the data ) , you must to pass a callback/promise or use async/await pattern after the function invocation.

So when you call :"user.save()", a promise should be provided, so this promise would tell you if there is an error in the onsave call.. It would be easier way to know if there is a problem.
You should look the [mongoosejs]( http://mongoosejs.com/docs/documents.html) documentation.

4. If you pass strings to a objectID field and if this string is valid, it should work, but if it does not work,
Have you try to map ur array of strings with objectid format to an array of objectid's ?
You should try to parse those strings to  object ids..

Something like : 

myArrayObjectIDStrings.map((x)=>{
return mongoose.Types.ObjectId(x);
})

Let me know your comments.

Regards

**Comment by RobJmz** [comment]
Mau! You are a genius!!!
Thanks a lot!

I kept the same formating that was autogenerated by the tool (I didn't add the promise)

I replaced the next line:
`.then((user) => user ? _.merge(user, body).save() : null)`

With the following:
```

.then((user) => {
      if(user){
        user = _.merge(user, body);
        user.wroups = user.wroups.map((wroup)=>{
          return mongoose.Types.ObjectId(wroup);
        });
        user.save();
        return user
      }else{
        return null
      }
    })
```

(If anyone else uses the following to save an array of objectIds, remember to import mongoose)

Not maping my array of strings was the problem, after mapping it to objectId, it started saving properly.
